File: backend/src/domain/card/card-installment.ts
import { generateFrenchSchedule } from '../amortization/french-amortization';
import { addMonths } from '../../shared/date/add-months';
import { roundMoney } from '../shared/money';

/**
 * Parametros de entrada para construir el cronograma de una compra diferida.
 */
export interface InstallmentScheduleInput {
  /** Capital total de la compra diferida (en pesos). */
  principal: number;
  /** Tasa mensual efectiva como fraccion decimal (ej. 0.02 para 2 % M.V.). */
  monthlyRate: number;
  /** Numero de cuotas en que se difiere la compra. */
  numberOfInstallments: number;
  /** Fecha de compra en formato YYYY-MM-DD; la cuota 1 vence un mes despues. */
  startDate: string;
}

/**
 * Una cuota del cronograma de compra diferida.
 */
export interface InstallmentItem {
  /** Numero de cuota (inicia en 1). */
  number: number;
  /** Fecha de vencimiento de la cuota en formato YYYY-MM-DD. */
  dueOn: string;
  /** Porcion de capital amortizado en esta cuota. */
  principal: number;
  /** Interes del periodo. */
  interest: number;
  /** Saldo de capital pendiente despues de esta cuota. */
  balance: number;
}

/**
 * Construye el cronograma cuota a cuota de una compra diferida en tarjeta de credito.
 *
 * Reutiliza el motor frances (`generateFrenchSchedule`) para calcular la
 * amortizacion y asigna la fecha de vencimiento de cada cuota segun:
 *   dueOn(n) = addMonths(startDate, n)
 * La suma de los capitales amortizados cierra exactamente en `principal`.
 *
 * @param input - Capital, tasa mensual, numero de cuotas y fecha de compra.
 * @returns Arreglo de cuotas con numero, fecha de vencimiento, principal, interes y saldo.
 */
export function buildInstallmentSchedule(
  input: InstallmentScheduleInput,
): InstallmentItem[] {
  const { principal, monthlyRate, numberOfInstallments, startDate } = input;

  const schedule = generateFrenchSchedule({
    principal,
    monthlyRate,
    numberOfInstallments,
  });

  // Mapea cada fila del motor frances a la estructura de cuota diferida.
  // La cuota N vence addMonths(startDate, N) — la primera un mes despues de la compra.
  return schedule.rows.map((row) => ({
    number: row.number,
    dueOn: addMonths(startDate, row.number),
    principal: roundMoney(row.principal),
    interest: roundMoney(row.interest),
    balance: roundMoney(row.balance),
  }));
}

File: app/lib/features/budget/domain/entities/card_installment.dart
/// Item (cuota) individual de un plan diferido de tarjeta de credito.
class CardInstallmentItem {
  const CardInstallmentItem({
    required this.number,
    required this.dueOn,
    required this.principal,
    required this.interest,
    required this.balance,
  });

  /// Numero secuencial de la cuota.
  final int number;

  /// Fecha de vencimiento de la cuota (YYYY-MM-DD).
  final String dueOn;

  /// Capital amortizado en la cuota.
  final double principal;

  /// Interes causado en la cuota.
  final double interest;

  /// Saldo pendiente tras la cuota.
  final double balance;
}

/// Plan diferido de tarjeta de credito con su cronograma de cuotas.
class CardInstallmentPlan {
  const CardInstallmentPlan({
    required this.id,
    required this.accountId,
    required this.principal,
    required this.numberOfInstallments,
    required this.monthlyRate,
    required this.startDate,
    required this.status,
    required this.items,
    this.description,
  });

  /// UUID del plan.
  final String id;

  /// UUID de la tarjeta.
  final String accountId;

  /// Descripcion de la compra diferida.
  final String? description;

  /// Capital total diferido en pesos.
  final double principal;

  /// Numero total de cuotas.
  final int numberOfInstallments;

  /// Tasa mensual vencida del diferido (fraccion decimal).
  final double monthlyRate;

  /// Fecha de inicio del plan (YYYY-MM-DD).
  final String startDate;

  /// Estado del plan: 'active', 'completed', etc.
  final String status;

  /// Cronograma de cuotas.
  final List<CardInstallmentItem> items;
}
